cli/azd/pkg/prompt/prompter.go

```diff
@@ -111,6 +112,47 @@ func (p *DefaultPrompter) PromptLocation(
 	return loc, nil
 }
 
+// PromptResource uses the console (or external) prompter to allow the user to select a resource
+// of optional type `resourceType`. The selected resource Name will be returned.
+// The Name can be used with the ARM or Bicep template function `reference` or in an existing resource template's name
+// to get provisioned state data from a resource, or passed to the `resourceId` function to get the full resource ID
+// if you know the `resourceType`.
+func (p *DefaultPrompter) PromptResource(
+	ctx context.Context,
+	subId string,
+	msg string,
+	resourceType string,
+) (string, error) {
+	options := azapi.ListResourcesOptions{
+		ResourceType: resourceType,
+	}
+	resources, err := p.resourceService.ListResources(ctx, p.env.GetSubscriptionId(), &options)
+	if err != nil {
+		return "", fmt.Errorf("listing resources: %w", err)
+	}
+
+	slices.SortFunc(resources, func(a, b *azapi.Resource) int {
+		return strings.Compare(a.Name, b.Name)
+	})
+
+	// TODO: Add `optional` field to allow something like "Create a new resource" (similar to resources groups below) and return ""?
+	choices := make([]string, len(resources))
+	for idx, resource := range resources {
+		// TODO: Get location display names from account manager instead?
+		choices[idx] = fmt.Sprintf("%d. %s (%s)", idx+1, resource.Name, resource.Location)
+	}
+
+	choice, err := p.console.Select(ctx, input.ConsoleOptions{
+		Message: msg,
+		Options: choices,
+	})
```

Comment:
I think we should consider/handle if `choices` is empty.  Would that mean:  The template I want to deploy requires me to have an existing resource of type `resourceType` but I don't have any in my Subscription.  Probably returning an error before the prompt might help. Like: `Template requires an existing resource of type <fooo> but there none was found in the subscription`.

Reply:
Good point. I'll handle the requirement in the prompt instead of the prompter, leaving the prompter more generalized. The prompt error will have access to more information as well such as the parameter name (key).